File: src/main/java/br/com/beleza/repository/ProductRepositoryImpl.java
package br.com.beleza.repository;

import java.util.HashMap;
import java.util.Map;

import org.springframework.stereotype.Repository;

import br.com.beleza.model.Product;

@Repository
public class ProductRepositoryImpl implements ProductRepository {

	private Map<Integer, Product> products = new HashMap<>();
	
	@Override
	public void save(Product product) {
		if(products.containsKey(product.getSku())) {
			throw new ProductAlreadyExistException("Dois produtos são considerados iguais se os seus skus forem iguais");
		}
		products.put(product.getSku(), product);
	}

	@Override
	public void update(Product product) {
		products.replace(product.getSku(), product);
	}

	@Override
	public void delete(Product product) {
		products.remove(product.getSku());
	}

	@Override
	public Product find(Integer sku) {
		
		Product p = products.get(sku);
		
		if(p == null) {
			throw new ProductNotFoundException("Product not found "+sku);
		}
		return p;
	}
}

File: src/main/kotlin/com/example/testkotlin/repository/ProductRepositoryImpl.kt
package com.example.testkotlin.repository

import com.example.testkotlin.model.Inventory
import com.example.testkotlin.model.Product
import com.example.testkotlin.model.WareHouseType
import com.example.testkotlin.model.Warehouse
import org.springframework.stereotype.Repository

@Repository
class ProductRepositoryImpl:ProductRepository {

    companion object {
        val products = mutableListOf<Product>()
    }

    override fun findAll(): List<Product> = products

    override fun findBySku(sku: Int): Product? = products.find { it.sku == sku }

    override fun save(p: Product): Boolean = products.add(p)

    override fun update(p: Product): Boolean {
        var product = findBySku(p.sku)

        if (product != null) {
            product.apply {
                this.name = p.name
                this.inventory.warehouses.map { warehouse -> Warehouse(
                        warehouse.locality,
                        warehouse.quantity,
                        warehouse.type) }
            }
            return true
        }
        return false
    }

    override fun delete(sku: Int): Boolean = products.removeIf { it.sku == sku }

}